Layouts and Rendering in Rails > 2. Creating Responses (part 2) (#225)
campuscode/rails-guides-pt-BR: pt-BR/layouts_and_rendering.md

TIP: `send_file` is often a faster and better option if a layout isn't required.

#### Opções para `render`

> As chamadas para o método `render` geralmente aceitam cinco opções:

Reviewer: ```suggestion
As chamadas para o método `render` geralmente aceitam seis opções:
```
Opa, acho que encontramos um erro na doc original: são seis opções no total.

|                     | 507                        | :insufficient_storage            |
|                     | 508                        | :loop_detected                   |
|                     | 510                        | :not_extended                    |
|                     | 511                        | :network_authentication_required |

NOTE: Se você tentar renderizar um conteúdo junto com um código de status que não tem conteúdo
(100-199, 204, 205 ou 304), o contéudo será descartado da resposta.

##### A opção `:formats`

O Rails usa o formato especificado na solicitação (ou `:html` por padrão). Você pode
mudar isso passando a opção `:formats` com um símbolo ou um _array_:

```ruby
render formats: :xml
render formats: [:json, :xml]
```

Se um template com o formato especificado não existir, será gerado um erro `ActionView::MissingTemplate`.

##### A opção `:variants`

Isso diz ao Rails para procurar variações de template do mesmo formato.
Você pode especificar uma lista de variações passando a opção `:variants` com um símbolo ou um _array_.

Um exemplo de uso seria este.

```ruby
# called in HomeController#index
render variants: [:mobile, :desktop]
```

> Com esse conjunto de variantes, o Rails procurará o seguinte conjunto de modelos e usará o primeiro que existe.

Reviewer 2: ```suggestion
Com esse conjunto de variantes, o Rails procurará o conjunto de modelos a seguir e usará o primeiro que encontrar.
```

Só uma sugestão mais de estilo de escrita, porque me parece mais claro.

There are no items in this list <em>yet</em>.
```

#### Como evitar erros de renderização dupla

> Mais cedo ou mais tarde, a maioria dos desenvolvedores Rails verá a mensagem de erro "Só pode renderizar ou redirecionar uma vez por ação" ("Can only render or redirect once per action"). Embora isso seja irritante, é relativamente fácil de corrigir. Geralmente isso ocorre devido a um mal-entendido sobre o modo como o `render` funciona.

Reviewer: ```suggestion
Mais cedo ou mais tarde, a maioria das pessoas desenvolvedoras Rails verão a mensagem de erro "Só pode renderizar ou redirecionar uma vez por ação" ("Can only render or redirect once per action"). Embora isso seja irritante, é relativamente fácil de corrigir. Geralmente isso ocorre devido a um mal-entendido sobre o modo como o `render` funciona.
```

end
```

Certifique-se de usar `and return` em vez de `&& return`, porque `&& return` não funcionará devido à precedência do operador na linguagem Ruby.

> Observe que a renderização implícita feita pelo ActionController detecta se `render` foi chamado, portanto, o seguinte funcionará sem erros:

Reviewer: ```suggestion
Observe que a renderização implícita feita pelo _ActionController_ detecta se `render` foi chamado, portanto, o seguinte código funcionará sem erros:
```
